flytekit/interfaces/data/gcs/gcs_proxy.py

```diff
@@ -0,0 +1,112 @@
+from flytekit.tools import subprocess as _subprocess
+from flytekit.common.exceptions.user import FlyteUserException as _FlyteUserException
+
+
+if _sys.version_info >= (3,):
+    from shutil import which as _which
+else:
+    from distutils.spawn import find_executable as _which
+
+
+def _update_cmd_config_and_execute(cmd):
+    env = _os.environ.copy()
+    return _subprocess.check_call(cmd, env=env)
+
+
+class GCSProxy(_common_data.DataProxy):
+    _GS_UTIL_CLI = "gsutil"
+
+    @staticmethod
+    def _check_binary():
+        """
+        Make sure that the AWS cli is present
+        """
+        if not _which(GCSProxy._GS_UTIL_CLI):
+            raise _FlyteUserException('gsutil (gcloud cli) not found at Please install.')
+
+    def exists(self, remote_path):
+        """
+        :param Text remote_path: remote gs:// path
+        :rtype bool: whether the gs file exists or not
+        """
+        GCSProxy._check_binary()
+
+        if not remote_path.startswith("gs://"):
+            raise ValueError("Not an GS Key. Please use FQN (GS ARN) of the format gs://...")
+
+        cmd = [GCSProxy._GS_UTIL_CLI, "-q", "stat", remote_path]
+        try:
+            _update_cmd_config_and_execute(cmd)
+            return True
+        except Exception:
+            return False
+
+    def download_directory(self, remote_path, local_path):
+        """
+        :param Text remote_path: remote s3:// path
+        :param Text local_path: directory to copy to
+        """
+        GCSProxy._check_binary()
+
+        if not remote_path.startswith("gs://"):
+            raise ValueError("Not an GS Key. Please use FQN (GS ARN) of the format gs://...")
+
+        cmd = [GCSProxy._GS_UTIL_CLI, "cp", "-r", remote_path, local_path]
+        return _update_cmd_config_and_execute(cmd)
+
+    def download(self, remote_path, local_path):
+        """
+        :param Text remote_path: remote s3:// path
+        :param Text local_path: directory to copy to
+        """
+        if not remote_path.startswith("gs://"):
+            raise ValueError("Not an GS Key. Please use FQN (GS ARN) of the format gs://...")
+
+        GCSProxy._check_binary()
+        cmd = [GCSProxy._GS_UTIL_CLI, "cp", remote_path, local_path]
+        return _update_cmd_config_and_execute(cmd)
+
+    def upload(self, file_path, to_path):
+        """
+        :param Text file_path:
+        :param Text to_path:
+        """
+        GCSProxy._check_binary()
+
+        cmd = [GCSProxy._GS_UTIL_CLI, "cp", file_path, to_path]
+
+        return _update_cmd_config_and_execute(cmd)
+
+    def upload_directory(self, local_path, remote_path):
+        """
+        :param Text local_path:
+        :param Text remote_path:
+        """
+        if not remote_path.startswith("gs://"):
+            raise ValueError("Not an GS Key. Please use FQN (GS ARN) of the format gs://...")
+
+        GCSProxy._check_binary()
+        cmd = [GCSProxy._GS_UTIL_CLI, "cp", "-r", local_path, remote_path]
```

Comment:
I think we can change `local_path` to `local_path + '\*' to just avoid upload the whole `engine_dir` folder.

Reply:
The current behaviour in this PR seems fine to me, because the expectation of this function is to upload the specified local path. Maybe it is the caller who should use `../*` instead?

Reply:
We can confirm that `gsutil` and `aws s3` behave differently when copying a dir recursively.

e.g:

```
$ tree test
test
└── test1
    └── test.txt

1 directories, 1 file

$ gsutil cp -r test/* gs://flyte-test
Copying file://test/test1/a/b/test.txt [Content-Type=text/plain]...
/ [1 files][    0.0 B/    0.0 B]
Operation completed over 1 objects.

$ gsutil ls gs://flyte-test
gs://flyte-test//
gs://flyte-test/test1/    <--------
```
vs.
```
$ gsutil cp -r test gs://flyte-test
Copying file://test/test1/a/b/test.txt [Content-Type=text/plain]...
/ [1 files][    0.0 B/    0.0 B]
Operation completed over 1 objects.

$ gsutil ls gs://flyte-test
gs://flyte-test//
gs://flyte-test/test/    <--------
```